fastedge-rust-sdk/examples/watermark/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+        Err(_) => {
+            return Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(Body::from("App misconfigured\n"))
+        }
+        Ok((u, h)) => (u, h),
+    };
+
+    /* Actual request to S3 */
+    let s3_req = Request::builder()
+        .method(Method::GET)
+        .uri(signed_url.as_str())
+        .header("Host", host)
+        .body(Body::empty())
+        .unwrap();
+    let rsp = match fastedge::send_request(s3_req) {
+        Err(_) => {
+            return Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(Body::empty())
+        }
+        Ok(r) => r,
+    };
+
+    // if response is not 200, just forward it to the caller
+    let (parts, body) = rsp.into_parts();
+    if parts.status != StatusCode::OK {
+        return Ok(Response::from_parts(parts, body))
+        // if you don't want to expose S3 error to the caller, just use
+        // return Response::builder()
+        //     .status(StatusCode::INTERNAL_SERVER_ERROR)
+        //     .body(Body::empty())
+
+    }
+
+    // load response as image
+    let buf = body.to_vec();
+    let img = match load_from_memory(buf.as_slice()) {
+        Ok(i) => i,
+        Err(_e) =>
+            // response body is not a valid image, just return it to the caller without changes
+            return Ok(Response::from_parts(parts, body))
+    };
+
+    // load watermark as image
+    let wm_img = match load_from_memory(wm_buf.as_slice()) {
+        Ok(i) => i,
+        Err(_e) =>
+            // should never happen
+            return Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(Body::from("Invalid watermark format\n"))
+    };
+
+    // get opacity from env
+    let opacity = match env::var("OPACITY").ok() {
+        None => DEFAULT_OPACITY,
+        Some(l) => match l.parse::<f32>() {
+            Err(_) => return Response::builder()    // opacity is not a number
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(Body::from("Invalid opacity value\n")),
+            Ok(v) => {
+                if v < 0.0 || v > 1.0 {     // opacity is not in 0-1.0 range
+                    return Response::builder()
+                        .status(StatusCode::INTERNAL_SERVER_ERROR)
+                        .body(Body::from("Invalid opacity value\n"))
+                }
+                v
+            }
+        },
+    }; 
+
+    let result = watermark(
+        &img,
+        &wm_img,
+        0,  // X offset for watermark placement
+        0,  // Y offset for watermark placement
+        opacity);
+
+    // convert resulting image to PNG
+    let mut c = Cursor::new(Vec::new());
+    let _ = result.write_to(&mut c, ImageOutputFormat::Png);
+    c.seek(SeekFrom::Start(0)).unwrap();
+    let mut out = Vec::new();
+    c.read_to_end(&mut out).unwrap();
```

Comment:
Here I see 2 allocation of vector and copy of the image bytes from one to other.
Could you test if this code can be optimised to:
`
let mut out = Vec::new();
let mut c = Cursor::new(out);
let _ = result.write_to(&mut c, ImageOutputFormat::Png);
`

Reply:
Even so, DynamicImage has a method into_bytes() to return Vec<u8>. But, you need to convert image to desired format first.

Reply:
Why output image is not save in the same format as input?

Reply:
yes, simplified this bit, also changed to use the same format